aptos-move/framework/src/natives/transaction_context.rs

```diff
@@ -34,6 +40,43 @@ impl NativeTransactionContext {
     }
 }
 
+/***************************************************************************************************
+ * native fun create_uuid
+ *
+ *   gas cost: base_cost
+ *
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct CreateUUIDGasParameters {
+    pub base: InternalGas,
+}
+
+fn native_create_uuid(
+    gas_params: &CreateUUIDGasParameters,
+    context: &mut SafeNativeContext,
+    mut _ty_args: Vec<Type>,
+    _args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    context.charge(gas_params.base)?;
+
+    let mut transaction_context = context
+        .extensions_mut()
+        .get_mut::<NativeTransactionContext>();
+    transaction_context.uuid_counter += 1;
+    const OBJECT_FROM_TRANSACTION_UUID_ADDRESS_SCHEME: u8 = 0xFB;
```

Comment:
since uuid's generated here are fully unique, we don't need to namespace them for each usage. 

but if we have multiple namespace-ing already, maybe we need a way to namespace it from all other namespaces? @movekevin  Is that what you meant here?

Reply:
We should also keep namespaces separate for different use cases. If the uuid generated from tx hash can be used in general use cases beyond object creation, we should also make sure that object creation uses its own domain separation. Otherwise, if in the future we ever support creating a different type of objects or entities (e.g. accounts, etc.) using these uuids, there would be conflicts.

cc @alinush as well regarding general domain separation best practice

Reply:
I am not sure I followed here

are you saying that native function create_uuid should take in domain/or we should have multiple create_uuid functions (create_uuid_for_object, create_uuid_for_account, etc) ?

All of that seems unnecessary. Code makes sure that create_uuid calls don't conflict with one another (i.e. if you call create_uuid multiple times in a transaction, from different places, you are going to get a different uuid back). as long as the place we plan to use it for didn't already use conflicting (i.e. identical namespace) there before, we should be good?

Reply:
There can code that uses the returned uuid directly as an address. In general, I think it's more cautious to always both have built in domain separation to prevent any possible conflicts with addresses generated from other places (e.g. resource account, other object creation functions) and also introduce namespacing in important paths where entities such as accounts or objects are created. Specifically:
1. Add a generic suffix here in uuid creation to make sure it never conflicts with any other way of generating addresses (e.g. resource account, object, etc.)
2. Add a suffix in object creation using the uuid. This prevents any conflicts with other functions that could reuse that same address to create accounts, for example.

Reply:
@movekevin - I still don't get it :D 

can you explain how this can make object generation and account creation conflict if they use this create_uuid utility?

Reply:
Anyone can create an account at a given address with aptos_account::create_account which does not add any extra domain separator. So if the domain separation is done inside create_guid and object::create_object doesn't add any domain separator, you can actually create an account and an object at the same address. Therefore, we should at least move the domain separation into the object creation function. We can additionally add another domain separator in create_guid, but that could be a bit paranoid as it'd be hard to get the same hex from other forms of generation (cc @alinush on that thought though)